discojs/src/processing/image.ts

```diff
@@ -0,0 +1,135 @@
+import { Repeat, Seq } from "immutable";
+import { createJimp } from "@jimp/core";
+import * as jimpResize from "@jimp/plugin-resize";
+
+import { Image } from "../index.js";
+
+/** Image where intensity is represented in the range 0..1 */
+export class NormalizedImage<
+  D extends 1 | 3 | 4 = 1 | 3 | 4,
+  W extends number = number,
+  H extends number = number,
+> {
+  // private as it doesn't check that array content is valid
+  private constructor(
+    public readonly data: Readonly<Float32Array>,
+    public readonly width: W,
+    public readonly height: H,
+    public readonly depth: D,
+  ) {
+    if (data.length != width * height * depth)
+      throw new Error("data isn't of excepted size");
+  }
+
+  static from<
+    D extends 1 | 3 | 4 = 1 | 3 | 4,
+    W extends number = number,
+    H extends number = number,
+  >(image: Image<D, W, H>): NormalizedImage<D, W, H> {
+    return new NormalizedImage(
+      Float32Array.from(image.data).map((v) => v / 255),
+      image.width,
+      image.height,
+      image.depth,
+    );
+  }
+}
+
+/** Add a full opaque alpha channel to an image */
+function addAlpha<W extends number, H extends number>(
+  image: Image<3 | 4, W, H>,
+): Image<4, W, H> {
+  switch (image.depth) {
+    case 3:
+      return new Image(
+        Uint8Array.from(
+          Seq(image.data).flatMap((v, i) => {
+            if (i % 3 !== 2) return [v];
+            else return [v, 0xff];
+          }),
+        ),
+        image.width,
+        image.height,
+        4,
+      );
+    case 4:
+      return new Image(image.data, image.width, image.height, image.depth);
+  }
+}
+
+/** Remove the alpha channel of an image */
+export function removeAlpha<W extends number, H extends number>(
+  image: Image<4, W, H>,
+): Image<3, W, H>;
+export function removeAlpha<
+  D extends 1 | 3,
+  W extends number,
+  H extends number,
+>(image: Image<D | 4, W, H>): Image<D, W, H>;
+export function removeAlpha<W extends number, H extends number>(
+  image: Image<1 | 3 | 4, W, H>,
+): Image<1 | 3, W, H> {
```

Comment:
correct, it's the [function overloads](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) syntax of typescript. it allows to specify more precisely how types get transformed by a function. in this case:
* `Image<4> -> Image<3>`
* `Image<1 | 3 | 4> -> Image<1 | 3>` (more exactly, it'll drop the 4 from the type)

and the last one is the actual implementation, which should behave how the overload are stating (but it is not actually enforced :roll_eyes:).

it couldn't have been done using a single function as otherwise, the type of the inputed image would get widden without reason. for example, `Image<3> -> Image<1 | 3>` (no alpha channel should keep the depth) & `Image<4> -> Image<1 | 3>` (but we know it's 3). it's a bit magic I agree :sparkles: but makes it very nice for the users.